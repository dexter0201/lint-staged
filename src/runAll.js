/* global process */
/* eslint no-console: 0 */

'use strict'

const sgf = require('staged-git-files')
const Listr = require('listr')
const stringifyObject = require('stringify-object')
const getConfig = require('./getConfig')
const runScript = require('./runScript')
const generateTasks = require('./generateTasks')
const resolveGitDir = require('./resolveGitDir')

module.exports = function runAll(packageJson, originalConfig) {
  const config = getConfig(originalConfig)
  const { gitDir, verbose, concurrent, renderer } = config
  sgf.cwd = resolveGitDir(gitDir)

  if (verbose) {
    console.log(`
Running lint-staged with the following config:
${stringifyObject(config)}

`)
  }

  return new Promise((resolve, reject) => {
    sgf('ACM', (err, files) => {
      if (err) {
        console.error(err)
      }

      /* files is an Object{ filename: String, status: String } */
      const filenames = files.map(file => file.filename)
      const tasks = generateTasks(config, filenames).map(task => ({
        title: `Running tasks for ${task.pattern}`,
        task: () =>
          new Listr(runScript(task.commands, task.fileList, packageJson, config), {
            // In sub-tasks we don't want to run concurrently
            // and we want to abort on errors
            concurrent: false,
            exitOnError: true
          }),
        skip: () => {
          if (task.fileList.length === 0) {
            return `No staged files match ${task.pattern}`
          }
          return false
        }
      }))

      if (tasks.length) {
        new Listr(
          tasks,
          Object.assign({}, config, {
            concurrent,
            renderer,
            exitOnError: !concurrent // Wait for all errors when running concurrently
          })
        )
          .run()
          .then(() => {
            resolve()
          })
          .catch(error => {
            if (Array.isArray(error.errors)) {
              error.errors.forEach(lintError => {
                console.error(lintError.message)
              })
            } else {
              console.log(error.message)
            }
            process.exit(1)
            reject()
          })
      }
    })
  })
}
