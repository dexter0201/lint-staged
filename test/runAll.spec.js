/* eslint no-console: 0 */

import stripColors from 'strip-ansi'
import runAll from '../src/runAll'

let stdout = ''
console.log = jest.fn(input => {
  // eslint-disable-next-line prefer-template
  stdout += stripColors(input).replace(/\[\d\d:\d\d:\d\d\]\W/, '') + '\n'
  return stdout
})

const packageJson = {
  scripts: {
    mytask: 'echo "Running task"'
  }
}

describe('runAll', () => {
  beforeEach(() => {
    process.stdout.isTTY = false // Overwrite TTY mode in order for Listr to use verbose renderer
    stdout = ''
  })

  it('should ouput config in verbose mode', done => {
    const config = {
      verbose: true,
      linters: {
        '*': 'mytask'
      }
    }

    runAll(packageJson, config).then(() => {
      expect(stdout).toMatchSnapshot()
      done()
    })
  })

  it('should not output config in non verbose mode', done => {
    const config = {
      '*': 'mytask'
    }

    runAll(packageJson, config).then(() => {
      expect(stdout).toMatchSnapshot()
      done()
    })
  })
})
