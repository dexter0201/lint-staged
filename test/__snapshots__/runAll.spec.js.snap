// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runAll should not output config in non verbose mode 1`] = `
"Running tasks for * [started]
Running tasks for * [skipped]
→ No staged files match *
"
`;

exports[`runAll should ouput config in verbose mode 1`] = `
"
Running lint-staged with the following config:
{
	verbose: true,
	linters: {
		'*': 'mytask'
	},
	concurrent: true,
	chunkSize: 9007199254740991,
	gitDir: '.',
	globOptions: {
		matchBase: true,
		dot: true
	},
	subTaskConcurrency: 1,
	renderer: 'verbose'
}


Running tasks for * [started]
Running tasks for * [skipped]
→ No staged files match *
"
`;
